### ISSUE TITLE
Difficulty reproducing VPT inference validation loss with provided weights/data/code. Possible bug in behavioural_cloning.py?

### ISSUE THREAD

**Issue by GaPala** [openai/Video-Pre-Training#29]
The VPT paper achieves a loss of ~2.5 on the contractor dataset using the 3x width foundation model. 

I am trying to reproduce this measurement (no training, just measuring the performance of the given models on the given data using the given code) by running inference with the provided 3x foundation weights over ~30 of the provided contractor videos (split between the Obtain Diamond task and House Building from Scratch task), using behavioural_cloning.py and commenting out the lines for weight updates so it just prints out the loss as it goes through the videos.

From my understanding, this inference pipeline should work similarly to how it is done in the paper, and should reproduce a similar  loss of around ~2.5, but the loss I measure is steadily around ~13 using IDM labels (using a custom IDM labelling script with the 64 window size in the centre, as suggested in the paper), and ~15 using the contractor labels.

If I allow the model to train, the loss does decrease quickly to ~3, but it behaves erratically when running with run_agent.py. I suspect that the pipeline/loss in behavioral cloning.py is somehow different from the original paper in a way that does not perform well in MineRL.

I edited behavioral_cloning.py to print out how often the model’s predicted action matches the ground truth, and it does increase greatly with training - but it doesn't make sense to me that genuine performance would improve so much from a loss that high with just a little training (~500 batches of batch size 8 at default settings), I suspect the model was trained in a significantly different pipeline in the paper, and that behavioural_cloning.py has a slightly different loss function, causing an initially high loss to be measured, but during finetuning the model quickly adapts to the new loss function, possibly because it is a simple transform of the original.

The main issue, behaviourally, is that even small amounts of training cause the model’s outputs to collapse to the most common outputs - just button index 4 and camera index 60. Increasing the batch size to 16 and the number of workers to 30 (using the 2x weights) does not improve the issue. Accounting for the class imbalance by proportionally removing overly-common class labels from contributing to the loss function decreases the loss to around 5, but the resulting behaviour is still erratic, so I don't think it's the root cause.

Thank you very much for any help with this! (And thanks to the original devs! This work is awesome)

**Comment by GaPala** [comment]
Hi, thanks for the reponse! I wanted to clarify that I'm not attempting to train the model; I understand that the given code does not implement backpropogation through time or other complexities that would have been in the original paper.

I am just trying to take the 3x foundation model, check its log loss over the given contractor data (plain inference), and expect the result to measure ~around what the paper reports as their final training loss on the contractor data. Since we are provided with the same models and same contractor data and similar code, it should be possible to just run inference with the models over the data and verify that they do get that level of performance on that data. My issue is that I can't replicate that.

So I don't think issues like sampling uniformity or the sequence length during the forward pass are a concern, since they  shouldn't affect inference performance.

I agree we can't expect to be able to train the models as well as the paper can, this is why I'm confused by how, when I run behavioural_cloning.py with training enabled (I am usually running it with training disabled, I enable training just to investigate the issue here), the model very quickly becomes much more accurate at predicting the actions than when it starts out (from ~13 loss to ~3 with a few minutes of training on a single GPU. 13 loss is actually only visible if you disable training, the first loss printed with training enabled is around 5) - this doesnt make sense to me because the original model should be strong enough that it is somewhat difficult to train it to become better at predicting actions, so I suspect something is causing the training setup to be out of distribution for the model, causing an initially high loss, but because it can take advantage of its pretraining it can quickly adapt to the new, different loss. 

The above normally wouldnt be too much of an issue, since the end loss is still low regardless, but there is another (imo  related) issue: from my understanding, improving the loss should result in more intelligent behaviour when running the agent with run_agent.py, but while even small amounts of finetuning with behavioural_cloning.py greatly improve the loss, they actually cause the agent to behave erratically in run_agent.py. 

I think the loss function has some bug which incentivises erratic behaviour, which explains why I can't replicate the original loss in the paper and why finetuned agents become erratic.

I think it is some issue like the last point you brought up. I understand there must be some differences in how data is handled and other details, but from my understanding these should be fairly minor details that still result in overall comparable behaviour to the paper, but from my experimenting I think the differences are actually large, so I suspect there is some subtle issue in the code, either with how the loss function is calculated or how actions are transformed between the agent and MineRL formats.

Or maybe I am just misunderstanding something? Sorry to drag this out further, hopefully this at least highlights the cause of my confusion!

**Comment by Miffyli** [comment]
Ah, thanks for clarifying your point! I skipped the "initial high loss part completely" ^^.

In short: I too am bit baffled by this and tried my best to fix the issues, but ultimately wrote it off as "something tiny small but important is off". The especially curious thing is that rolling out existing models results in same results as reported in the paper, so the data preprocessing steps should be right.

The most likely reason for high initial loss I can think of is that the loss was different during training and/or the way it is reported in the paper was different. Unfortunately we were not able to include any of the training details from the original code here, so it is unknown if they had some special steps in the loss calculation (e.g., weights, some tuning). Some of the preprocessing steps could be off, e.g., paper says they used no-op filtering. This code also ignores no-op actions, but it unclear if they skip the whole step (i.e. do not see the frame at all) or just ignore the action in training. In this code we ignore the whole frame, which is lazier way of fixing it but potentially wrong.


The poor performance with finetuning stems from the above thing: since the finetuning is out-of-distribution of the original model, even tiny updates will quickly take model off from the original behaviour. However this might be just result of the single-step training as well (or the poor data-loading). 

Two things you could try for better finetuning results:
- Check out the [BASALT 2022 BC baseline code](https://github.com/minerllabs/basalt-2022-behavioural-cloning-baseline/blob/main/behavioural_cloning.py). It has more refined BC implementation with e.g., KL-div loss w.r.t original model to avoid completely breaking
- Try using learning rate warmup, where for first 100-500 updates you linearly anneal the learning rate from 0 to `initial_lr`. This is apparently a thing transformer training needs, and is not implemented here. It might be one of the things that also break training.
- Try increasing gradient accumulation to very large numbers (e.g., 1024)

**Comment by GaPala** [comment]
Thank you very much for your help! I think I found the issues!

1. **I think the provided contractor mp4 videos are somehow out of distribution**

I collected some Minecraft videos online and ran the IDM over them to get labels. Running the VPT model on this dataset matches the reported loss in the paper of around 2-3! Since this reproduces the paper results I'm very happy with this and I'll mark this issue as closed.

I briefly tried experimenting with training the VPT model on the provided contractor data but with everything frozen except the CNN block, and the loss still went down fairly quickly (though not as fast as when training the whole model), so I suspect the difference is something to do with the contractor videos being OOD.

2. **I think the suggested learning rate was too high**

I havent experimented too much with this, but during training, the loss would drop much more quickly and reach a lower value if I used a lower loss (I quickly experimented with 0.0000181 and grad_norm=1.0)


Thank you very much for your support with this!
